libcaf_net/caf/net/http/client.cpp

```diff
@@ -0,0 +1,186 @@
+  auto& buf = down_->output_buffer();
+  auto size_str = detail::format("{:X}\r\n", bytes.size());
+  std::transform(size_str.begin(), size_str.end(), std::back_inserter(buf),
+                 [](auto c) { return static_cast<std::byte>(c); });
+  buf.insert(buf.end(), bytes.begin(), bytes.end());
+  buf.emplace_back(std::byte{'\r'});
+  buf.emplace_back(std::byte{'\n'});
+  return down_->end_output();
+}
+
+bool client::send_end_of_chunks() {
+  std::string_view str = "0\r\n\r\n";
+  auto bytes = as_bytes(make_span(str));
+  down_->begin_output();
+  auto& buf = down_->output_buffer();
+  buf.insert(buf.end(), bytes.begin(), bytes.end());
+  return down_->end_output();
+}
+
+void client::switch_protocol(std::unique_ptr<octet_stream::upper_layer> next) {
+  down_->switch_protocol(std::move(next));
+}
+
+// -- octet_stream::upper_layer implementation ---------------------------------
+
+error client::start(octet_stream::lower_layer* down) {
+  down_ = down;
+  return up_->start(this);
+}
+
+void client::abort(const error& reason) {
+  up_->abort(reason);
+}
+
+void client::prepare_send() {
+  up_->prepare_send();
+}
+
+bool client::done_sending() {
+  return up_->done_sending();
+}
+
+ptrdiff_t client::consume(byte_span input, byte_span) {
+  CAF_LOG_TRACE(CAF_ARG2("bytes", input.size()));
+  ptrdiff_t consumed = 0;
+  if (mode_ == mode::read_header) {
+    if (input.size() >= max_response_size_) {
+      abort("Header exceeds maximum size.");
+      return -1;
+    }
+    auto [hdr, remainder] = v1::split_header(input);
+    // Wait for more data.
+    if (hdr.empty())
+      return 0;
+    // Note: handle_header already calls up_->abort().
+    if (!handle_header(hdr))
+      return -1;
+    // Prepare for next loop iteration.
+    consumed = static_cast<ptrdiff_t>(hdr.size());
+    input = remainder;
+    // Transition to the next mode.
+    if (hdr_.chunked_transfer_encoding()) {
+      mode_ = mode::read_chunks;
+    } else if (auto len = hdr_.content_length()) {
+      // Protect against payloads that exceed the maximum size.
+      if (*len >= max_response_size_) {
+        abort("Payload exceeds maximum size.");
+        return -1;
+      }
+      // Transition to read_payload mode and continue.
+      payload_len_ = *len;
+      mode_ = mode::read_payload;
+    } else {
+      // TODO: we may *still* have a payload since HTTP can omit the
+      //       Content-Length field and simply close the connection
+      //       after the payload.
+      if (!invoke_upper_layer(const_byte_span{}))
+        return -1;
+      return consumed;
+    }
+  }
+  if (mode_ == mode::read_payload) {
+    if (input.size() >= payload_len_) {
+      if (!invoke_upper_layer(input.subspan(0, payload_len_)))
+        return -1;
+      consumed += static_cast<ptrdiff_t>(payload_len_);
+      input.subspan(payload_len_);
+      mode_ = mode::read_header;
+      return consumed;
+    } else {
+      // Wait for more data.
+      return consumed;
+    }
```

Comment:
I'd push this up to get rid of one layer of `if` nesting:

```
  if (input.size() >= payload_len_) {
    // Wait for more data.
    return 0;
  }
```

Reply:
Done.